### PR TITLE
Support for Logout to work with ADFS 3.0 as Identity Provider

### PR DESCRIPTION
**Pull request by mickael83**
The following changes were made to fix Logout . I tested it with ADFS 3.0.

1. I added some code to save in session the NameId and SessionIndex received through the acs url. I use this values in the LogoutRequest in order to logout to work correctly (without it ADFS does not know which is the session participant that is logging out)

2. For some reason, in sls process if the method processSLO is called with the parameter $retrieveParametersFromServer to false, the signature validation of the LogoutResponse fails, if called with true, the signature is correctly validated (I found out that the signedQuery generated by onelogin LogoutResponse class (line 153) differs depending on the $retrieveParametersFromServer parameter). Maybe it would be better to make this parameter configurable.

### CODE REVIEW

**Comment by aacotroneo** [comment]
sorry about the Session object, we are removing all facades.. please use `session()` helper

**Comment by aacotroneo** [comment]
I understand your problem, but so far this library didn't use any session at all... Using or not session will depend on each project and I wouldn't want to impose it as a restriction (ie. you could create a token when you decode the saml2 token and send it to your SPA.. you won't need a session)

Because of this, I ended up exposing the Login events with the user, so the porject can save it in session or not. May be we can add some events or other way to let you execute your session related code from within your project.

File: src/Aacotroneo/Saml2/Saml2Auth.php [hunk 1]
```diff
@@ -84,6 +85,11 @@ function acs()
             return $errors;
         }
 
+        Session::put('federationData', [
+            'sessionIndex' => $auth->getSessionIndex(),
+            'nameId' => $auth->getNameId()
+        ]);
```
Comment (aacotroneo) [inline src/Aacotroneo/Saml2/Saml2Auth.php]: This `auth` is passed to the User (which is latter exposed to the project). You could get this information from there (extend the user to make it available)

File: src/Aacotroneo/Saml2/Saml2Auth.php [hunk 2]
```diff
@@ -106,7 +112,8 @@ function sls()
             \Event::fire(new Saml2LogoutEvent());
         };
 
-        $auth->processSLO($keep_local_session, null, false, $session_callback);
+        // Mickael: Changed $retrieveParametersFromServer paramter to true otherwise when the signedQuery is corrupted and the signature validation failed
+        $auth->processSLO($keep_local_session, null, true, $session_callback);
```
Comment (aacotroneo) [inline src/Aacotroneo/Saml2/Saml2Auth.php]: we can move this to the configuration files... Im not sure about its implication to existing users.

File: src/Aacotroneo/Saml2/Saml2Auth.php [hunk 3]
```diff
@@ -59,11 +60,11 @@ function login($returnTo = null)
...
      * local session if applicable.
      */
-    function logout()
+    function logout($returnTo = null)
     {
         $auth = $this->auth;
-
-        $auth->logout();
+        $federationData = Session::get('federationData');
+        $auth->logout($returnTo, array(), $federationData['nameId'], $federationData['sessionIndex']);
```
Comment (aacotroneo) [inline src/Aacotroneo/Saml2/Saml2Auth.php]: if we could send the federation data as a parameter... we can make them null by default too, and we are done I think

**Comment by mickael83** [comment]
1. I'll try to implement your suggestions however I have a doubt on how to get the auth object from the Saml2User instance passed in Saml2LoginEvent to get the session index and the nameId. How can I extend the Saml2User class and tell laravel to use my custom implementation?

2. I will change the Saml2Controller class to check if nameId and sessionIndex parameters exists;

3. I will change the Saml2Auth logout method to pass optionally as parameter  the nameId and the sessionIndex;

4. I will add the configuration to choose to use or not the $retrieveParametersFromServer option.

**Comment by aacotroneo** [comment]
for 1- I meant to modify the source and create a PR :)

**Comment by mickael83** [comment]
Duplicated PR (#30)

**Pull request closed without merging**